### PR TITLE
Implement the Validate RPC on built-in plugins

### PR DESCRIPTION
**Pull request by edwbuck**
Update all plugins to have a Validate gRPC and centralize the existing validation logic such that it is reached from Validate and Configure gRPC calls.

### CODE REVIEW

File: pkg/server/plugin/upstreamauthority/awspca/pca.go
Comment (amartinezfayo) [inline pkg/server/plugin/upstreamauthority/awspca/pca.go:64]: Why don't have a `validate` function that has the validation logic only? It doesn't seem like creating a new config is needed when performing a validation. The returned config will always be discarded.
In the same way as this function, the `validate` function could be called from the Configure and Validate functions.

Comment (amartinezfayo) [reply pkg/server/plugin/upstreamauthority/awspca/pca.go:64]: The function does not need to be exported.

File: pkg/server/plugin/upstreamauthority/disk/disk.go
Comment (amartinezfayo) [inline pkg/server/plugin/upstreamauthority/disk/disk.go:139]: I think that we should probably check for an error here. If there is an error, I would expect the ValidateResponse to be nil.

File: pkg/server/plugin/upstreamauthority/vault/vault.go
Comment (amartinezfayo) [inline pkg/server/plugin/upstreamauthority/vault/vault.go:70]: I understand that this unusable bool is needed to have appended notes. I believe that current validations in SPIRE returns a single error, even if as a result of continuing with the validation it would return more errors, so it wouldn't be surprising to return a single note.
I'm making this comment just to point that we could simplify the logic just returning a single note, but I don't feel too strongly on it.

File: pkg/server/plugin/upstreamauthority/awspca/pca.go
Comment (edwbuck) [reply pkg/server/plugin/upstreamauthority/awspca/pca.go:64]: A plugin.Validate gRPC that is independent of the plugin.Configure gRPC would mean that plugin configuration validation would have to be cut-and-pasted twice into the Validate and Configure gRPC calls.  Rather than do that, both calls should defer to one config checking routine (as discussed and accepted by the maintainers).  By pushing that routine into the NewConfiguration func, all new configurations are now centralizing all of the configuration checking.

File: pkg/server/plugin/upstreamauthority/disk/disk.go
Comment (edwbuck) [reply pkg/server/plugin/upstreamauthority/disk/disk.go:139]: The error is checked and sent in the response.  Whether a validation succeeds or fails, the response is always sent successfully, because the response contains whether it succeeded or failed.

That was the design that was agreed upon by the maintainers.

File: pkg/server/plugin/upstreamauthority/vault/vault.go
Comment (edwbuck) [reply pkg/server/plugin/upstreamauthority/vault/vault.go:70]: There are only calls to configure a plugin, there are not current calls to validate a plugin's contents.

The maintainers insisted that for validation, all failures or successful notes be returned in a group, so one doesn't have to validate the plugin config multiple times to just get "the next error".  I agreed with this request, and it was worked into the plugin sdk.  Asking for a change like this is basically asking that the maintainers rescind their prior requests, when they were clearly improvements.  I don't think that's appropriate.

The code above was worked to ensure that the same single note that fails the Configure call is the single returned error under the Configure usage of NewConfiguration.  It also returns multiple notes (some of which might not be errors) in the Validate calls.  This permits one to validate once, getting a list of issues, notes, or errors.

File: pkg/server/plugin/upstreamauthority/awspca/pca.go
Comment (edwbuck) [reply pkg/server/plugin/upstreamauthority/awspca/pca.go:64]: @amartinezfayo The `Configuration` struct across most (if not all) plugins is also exported.  Would you like for me to change that at the same time I change `NewConfiguration`?

**Comment by azdagron** [comment]
Something like this might help establish a pattern and reduce boilerplate: https://gist.github.com/azdagron/b55f1e70e1855cfa1aa73cb8f0027699

File: pkg/server/plugin/upstreamauthority/awspca/pca.go
Comment (edwbuck) [reply pkg/server/plugin/upstreamauthority/awspca/pca.go]: @amartinezfayo In review with @azdagron a different name for the function was suggested, `newConfig`, which also is not exported.  That method is applied as a parameter to the new functor `pluginconf.Build` which is exported because it lies in a different module.

I'm closing this item as the method in the Plugin is no longer exported.  If you feel I'm closing this in error, please reopen it.

As for only validating the configuration, without creating a new `Configuration` object, such a thing cannot be done because the HCL parsing libraries require a `Configuration` object to parse the HCL out of the text string (now named `hclText`.  While it might be a point to argue where that `Configuration` object is created prior to the HashiCorp parsing of HCL1, there is no way to validate the configuration is HCL1 without parsing it, requiring the prototype, unfilled `Configuration` object.

After that object is filled with values as a side-effect of the parsing, additional validation is performed on the fields post-parsing for many of the plugins.  That validation clearly needs to happen in both the `Configure(...)` and `Validate(...)` gRPC calls, `Configure(...)` to ensure that it is checked before being used, and `Validate(...)` to ensure it is checked before being considered valid.  We cannot validate the values in `validation` logic only, unless we have the `Configure(...)` gRPC call `validate`; but, if we do that, then the return values of a pure-validation-only call won't return a `Configure` object and we will have to build the configuration object twice.

This solution, with the validation returning a `Configuration` object, the list of `notes`, and a possible `error` satisfies both calls from gRPCs `Configure(...)` and `Validate(...)` by letting `Configure(...)` discard the notes and `Validate(...)` discard the Configuration object.

File: pkg/server/plugin/upstreamauthority/vault/vault.go
Comment (edwbuck) [reply pkg/server/plugin/upstreamauthority/vault/vault.go]: @amartinezfayo @azdagron 

In the request to return multiple potential notes (including multiple potential errors) in one call to the gRPC `Validate(...)`, we made the need to return multiple errors, if they could be reported in one pass part of the solution's requirements.

**Comment by edwbuck** [comment]
> Something like this might help establish a pattern and reduce boilerplate: https://gist.github.com/azdagron/b55f1e70e1855cfa1aa73cb8f0027699

The pattern of coupling the configuration usability control and the note as an error or informational message under one API was adopted and applied to the plugins that have been reworked under this PR.

**Comment by amartinezfayo** [comment]
Hi @edwbuck, I spent some time looking at the aws_kms test failures.
I left some comments. In summary, I don't think that we need to make any change in the way that plugintest.Load is called in the setupTest function, so most of the changes can be avoided.

Here is a patch that essentially removes those changes and makes `go test -race` to pass.
[aws_kms-test-fix.patch](https://github.com/user-attachments/files/17003527/aws_kms-test-fix.patch)

**Comment by edwbuck** [comment]
> Hi @edwbuck, I spent some time looking at the aws_kms test failures. I left some comments. In summary, I don't think that we need to make any change in the way that plugintest.Load is called in the setupTest function, so most of the changes can be avoided.
> 
> Here is a patch that essentially removes those changes and makes `go test -race` to pass. [aws_kms-test-fix.patch](https://github.com/user-attachments/files/17003527/aws_kms-test-fix.patch)

Thank you Agustin.  I read over the patch, and in all other files I didn't notice that this combined two different testing patterns, such that it configured each plugin twice, creating the conditions that eventually triggered my race condition with my additional call to plugintest.CoreConfig(...).

Expect a push soon.

**Comment by edwbuck** [comment]
Ready for review two weeks ago.  Please prioritize so I can stop keeping this up-to-date with the moving main branch.

File: pkg/server/plugin/upstreamauthority/awssecret/awssecret.go
Comment (azdagron) [inline pkg/server/plugin/upstreamauthority/awssecret/awssecret.go:63]: nit: does not follow naming conventions:
```suggestion
	envSessionToken, envSessionTokenExists := os.LookupEnv("AWS_SESSION_TOKEN")
```

File: pkg/server/plugin/upstreamauthority/awssecret/awssecret.go
Comment (azdagron) [inline pkg/server/plugin/upstreamauthority/awssecret/awssecret.go:72]: Is this validation code new? I don't think it is correct either. This value is only used when attempting to use static credentials. Otherwise the AWS SDK can load credentials from many different sources.

In general, I don't think we should introduce behavior changes as part of this PR.

File: pkg/server/plugin/upstreamauthority/certmanager/certmanager.go
Comment (azdagron) [inline pkg/server/plugin/upstreamauthority/certmanager/certmanager.go:137]: This is also checked by pluginconf.Build, correct?

File: pkg/server/plugin/upstreamauthority/disk/disk.go
Comment (azdagron) [inline pkg/server/plugin/upstreamauthority/disk/disk.go:30]:
```suggestion
	CoreConfigTrustDomainRequired  = "server core configuration must contain trust_domain"
	CoreConfigTrustDomainMalformed = "server core configuration trust_domain is malformed"
```

File: pkg/agent/plugin/nodeattestor/tpmdevid/devid_test.go
Comment (edwbuck) [reply pkg/agent/plugin/nodeattestor/tpmdevid/devid_test.go:231]: I agree, it's a side-effect of having package `pluginconf` only return `InvalidArgument` status codes.

To fix this, I've moved the autodetection outside of the configurnation handling, handling it just prior to use of the new configuration in `Configure`.  It's triggered on the `Config`'s new `bool` field `Autodetect` which is false unless the conditions are right for autodetection.

File: pkg/agent/plugin/workloadattestor/docker/docker_posix.go
Comment (edwbuck) [reply pkg/agent/plugin/workloadattestor/docker/docker_posix.go:49]: Fixed in next update.

File: pkg/agent/plugin/workloadattestor/systemd/systemd_windows.go
Comment (edwbuck) [reply pkg/agent/plugin/workloadattestor/systemd/systemd_windows.go:39]: Fixed in next push (so I can sign it, accepting the signoff and commit messes up signatures occasionally).

File: pkg/agent/plugin/workloadattestor/unix/unix_windows.go
Comment (edwbuck) [reply pkg/agent/plugin/workloadattestor/unix/unix_windows.go:41]: Fixed in next push, so I can sign it.

File: pkg/server/plugin/upstreamauthority/awssecret/awssecret.go
Comment (edwbuck) [reply pkg/server/plugin/upstreamauthority/awssecret/awssecret.go:63]: Fixed in next push, so I can sign it.

File: pkg/server/plugin/upstreamauthority/awssecret/awssecret.go
Comment (edwbuck) [reply pkg/server/plugin/upstreamauthority/awssecret/awssecret.go:72]: In the old `validateConfig(...)` the security token was set like so

> if config.SecurityToken == "" {
>              config.SecurityToken = p.hooks.getenv("AWS_SESSION_TOKEN")
> }

And there were absolutely no errors if it was set in the config, the environment, both, or neither.

I'm restoring the no error checking behavior.

File: pkg/server/plugin/upstreamauthority/certmanager/certmanager.go
Comment (edwbuck) [reply pkg/server/plugin/upstreamauthority/certmanager/certmanager.go:137]: Yes, and I'm cutting in in the next push.

It was a remnant of the pre-`pluginconf.Build` approach.  Certmanager.go was one of the first plugins to be worked, and the `pluginconf.Build` came afterwards.

File: pkg/server/plugin/upstreamauthority/disk/disk.go
Comment (edwbuck) [reply pkg/server/plugin/upstreamauthority/disk/disk.go:30]: Fixed in next push, so I can sign it.

**Review by azdagron — APPROVED** [review]

**Pull request merged**
